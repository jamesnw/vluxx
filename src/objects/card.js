export default class Card {
  constructor(title, type, apply) {
    this.title = title;
    this.type = type;
    this.apply = apply;
  }
}
