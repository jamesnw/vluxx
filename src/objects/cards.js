const TYPES = {
  GOAL: "goal",
  RULE: "rule",
  KEEPER: "keeper"
};
export const cards = [
  {
    title: "Chocolate Chip Cookie",
    type: TYPES.GOAL,
    goal: ["Chocolate", "Cookie"]
  },
  { title: "Bedtime snack", type: TYPES.GOAL, goal: ["Milk", "Cookie"] },
  { title: "Bedtime snack 1", type: TYPES.GOAL, goal: ["Milk", "Cookie"] },
  { title: "Bedtime snack 2", type: TYPES.GOAL, goal: ["Milk", "Cookie"] },
  { title: "Bedtime snack 3", type: TYPES.GOAL, goal: ["Milk", "Cookie"] },
  { title: "Bedtime snack 4", type: TYPES.GOAL, goal: ["Milk", "Cookie"] },
  { title: "Bedtime snack 5", type: TYPES.GOAL, goal: ["Milk", "Cookie"] },

  { title: "Chocolate", type: TYPES.KEEPER },
  { title: "Cookie", type: TYPES.KEEPER },
  { title: "Milk", type: TYPES.KEEPER },
  { title: "Milk 1", type: TYPES.KEEPER },
  { title: "Milk2", type: TYPES.KEEPER },
  { title: "Milk3", type: TYPES.KEEPER },
  { title: "Milk4", type: TYPES.KEEPER },
  { title: "Milk5", type: TYPES.KEEPER },
  { title: "Milk6", type: TYPES.KEEPER },
  { title: "Milk7", type: TYPES.KEEPER },
  { title: "Milk8", type: TYPES.KEEPER },
  { title: "Milk9", type: TYPES.KEEPER },
  { title: "Milk0", type: TYPES.KEEPER },
  { title: "Milk12", type: TYPES.KEEPER },
  { title: "Milk23", type: TYPES.KEEPER },
  { title: "Milk24", type: TYPES.KEEPER },
  { title: "Milk35", type: TYPES.KEEPER },

  { title: "Draw 2", type: TYPES.RULE, rule: { category: "Draw", amount: 2 } },
  { title: "Draw 3", type: TYPES.RULE, rule: { category: "Draw", amount: 3 } },
  { title: "Draw 4", type: TYPES.RULE, rule: { category: "Draw", amount: 4 } },
  { title: "Draw 5", type: TYPES.RULE, rule: { category: "Draw", amount: 5 } },
  { title: "Play 2", type: TYPES.RULE, rule: { category: "Play", amount: 2 } },
  { title: "Play 3", type: TYPES.RULE, rule: { category: "Play", amount: 3 } },
  { title: "Play 4", type: TYPES.RULE, rule: { category: "Play", amount: 4 } },
  { title: "Play 5", type: TYPES.RULE, rule: { category: "Play", amount: 5 } }
];

export function applyCard(game, card) {
  switch (card.type) {
    case TYPES.GOAL:
      game.goal = card;
      break;
    case TYPES.KEEPER:
      var player = game.players[game.activePlayer];
      player.keepers = player.keepers || [];
      player.keepers.push(card);
      game.players[game.activePlayer] = player;
      break;
    case TYPES.RULE:
      var rule = card.rule;
      if (rule.category === "Draw") game.constantRules.draw = rule.amount;
      else if (rule.category === "Play") game.constantRules.play = rule.amount;
      break;

    default:
      break;
  }
  return game;
}
