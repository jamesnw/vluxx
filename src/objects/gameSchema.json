{
  "type": "object",
  "definitions": {
    "card": {
      "type": "object",
      "properties": {
        "title": { "type": "string" },
        "type": { "type": "string" }
      },
      "required": ["title", "type"]
    },
    "player": {
      "type": "object",
      "properties": {
        "hand": {
          "type": "array",
          "items": [
            {
              "$ref": "#/definitions/card"
            }
          ]
        },
        "name": {
          "type": "string"
        }
      },
      "required": ["hand", "name"]
    }
  },
  "properties": {
    "activePlayer": {
      "type": "integer"
    },
    "constantRules": {
      "type": "object",
      "properties": {
        "draw": {
          "type": "integer"
        },
        "play": {
          "type": "integer"
        }
      },
      "required": ["draw", "play"]
    },
    "deck": {
      "type": "array",
      "items": [
        {
          "$ref": "#/definitions/card"
        }
      ]
    },
    "discard": {
      "type": "array",
      "items": [
        {
          "$ref": "#/definitions/card"
        }
      ]
    },
    "name": {
      "type": "string"
    },
    "players": {
      "type": "array",
      "items": [{ "$ref": "#/definitions/player" }]
    },
    "rules": {
      "type": "array",
      "items": {}
    },
    "state": {
      "type": "string",
      "enum": ["started", "won"]
    },
    "turn": {
      "type": "object",
      "properties": {
        "playsRemaining": {
          "type": ["integer", "null"]
        }
      },
      "required": ["playsRemaining"]
    }
  },
  "required": [
    "activePlayer",
    "constantRules",
    "deck",
    "discard",
    "name",
    "players",
    "rules",
    "state",
    "turn"
  ]
}
