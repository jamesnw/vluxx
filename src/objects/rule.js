export default class Rule {
  constructor(name) {
    this.name = name;
  }
}
