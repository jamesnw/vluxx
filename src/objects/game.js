import { cards } from "./cards";
import { dealCard } from "./player";
import _ from "lodash";

export function newGame(name) {
  return {
    name: name,
    players: [],
    rules: []
  };
}

export function startGame(game) {
  game.deck = _.shuffle(cards);
  game.state = "started";
  game.constantRules = {
    draw: 1,
    play: 1
  };
  game.turn = {};
  // deal
  for (let index = 0; index < 3; index++) {
    game.players.forEach((player, index) => {
      dealCard(game, index);
    });
  }
  game.activePlayer = 0;
  return game;
}
