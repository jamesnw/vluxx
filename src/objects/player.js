import _ from "lodash";
import { applyCard } from "./cards";
import { gameIsValid } from "./game";

export function dealCard(game, index) {
  index = index === undefined ? game.activePlayer : index;
  const card = game.deck.shift();
  game.players[index].hand = game.players[index].hand || [];
  game.players[index].hand.push(card);
  gameIsValid(game);
  return game;
}

export function playCard(game, card) {
  game.discard = game.discard || [];
  game.discard.push(card);
  let hand = game.players[game.activePlayer].hand;
  _.remove(hand, { title: card.title });
  game.players[game.activePlayer].hand = hand;
  game.turn.playsRemaining = game.turn.playsRemaining - 1;
  game = applyCard(game, card);
  gameIsValid(game);
  return game;
}

export function startTurn(game) {
  const cardsToDraw = game.constantRules.draw;
  for (let index = 0; index < cardsToDraw; index++) {
    game = dealCard(game);
  }
  game.turn = {
    playsRemaining: game.constantRules.play
  };
  gameIsValid(game);
  return game;
}

export function endTurn(game) {
  game.turn = { playsRemaining: null };
  game.activePlayer = (game.activePlayer + 1) % game.players.length;
  gameIsValid(game);
  return game;
}
