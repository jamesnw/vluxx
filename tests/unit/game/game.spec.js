import { newGame, startGame } from "@/objects/game";
import _ from "lodash";
describe("game", () => {
  describe("newGame", () => {
    it("should create basic structure", () => {
      expect(newGame("foo")).toMatchSnapshot();
    });
    it("should add game name", () => {
      const nameToUse = "Foo";
      let { name } = newGame(nameToUse);
      expect(name).toEqual(nameToUse);
    });
  });
  describe("startGame", () => {
    it("should add cards", () => {
      let { deck } = startGame({ players: [] });
      expect(deck).toEqual(expect.any(Array));
    });
    it("should set initial static values", () => {
      let game = startGame({ players: [] });
      game = _.pick(game, ["state", "constantRules", "turn", "activePlayer"]);
      expect(game).toMatchSnapshot();
    });
  });
});
