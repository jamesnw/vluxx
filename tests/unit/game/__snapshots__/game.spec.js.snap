// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`game newGame should create basic structure 1`] = `
Object {
  "name": "foo",
  "players": Array [],
  "rules": Array [],
}
`;

exports[`game startGame should set initial static values 1`] = `
Object {
  "activePlayer": 0,
  "constantRules": Object {
    "draw": 1,
    "play": 1,
  },
  "state": "started",
  "turn": Object {},
}
`;
